pkg/audit/manager.go

```diff
@@ -519,16 +524,43 @@ func (am *Manager) reviewObjects(ctx context.Context, kind string, folderCount i
 			augmentedObj := target.AugmentedUnstructured{
 				Object:    *objFile,
 				Namespace: ns,
+				Source:    mutationtypes.SourceTypeOriginal,
 			}
 			resp, err := am.opa.Review(ctx, augmentedObj)
 			if err != nil {
 				am.log.Error(err, "Unable to review object from file", "fileName", fileName, "objNs", objNs)
 				continue
 			}
 
-			results := ToResults(&augmentedObj.Object, resp)
+			// Expand object and review any resultant resources
+			base := &mutationtypes.Mutable{
+				Object:    objFile,
+				Namespace: ns,
+				Username:  "",
+				Source:    mutationtypes.SourceTypeOriginal,
+			}
+			resultants, err := am.expansionSystem.Expand(base)
+			if err != nil {
+				am.log.Error(err, "Unable to expand object", "objName", objFile.GetName())
+				continue
+			}
+			for _, resultant := range resultants {
+				au := target.AugmentedUnstructured{
+					Object:    *resultant.Obj,
+					Namespace: ns,
+					Source:    mutationtypes.SourceTypeGenerated,
+				}
+				resultantResp, err := am.opa.Review(ctx, au)
```

Comment:
For audit, I just tested the following deployment with 0 replica, which gets blocked during validation. However, if the same deployment already exists in a cluster, I expect to see it in the constraint's list of violations, but I do not.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 0
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
        args:
        - "/bin/sh"
```
@davis-haba  can you PTAL?

Reply:
Do you have the constraint, template and TemplateExpansion you were using?

Reply:
```yaml
apiVersion: constraints.gatekeeper.sh/v1beta1
kind: K8sAllowedRepos
metadata:
  name: prod-repo-is-openpolicyagent
spec:
  match:
    kinds:
      - apiGroups: [""]
        kinds: ["Pod"]
    namespaces:
      - "default"
  parameters:
    repos:
      - "openpolicyagent"
```
```yaml
apiVersion: templates.gatekeeper.sh/v1
kind: ConstraintTemplate
metadata:
  name: k8sallowedrepos
spec:
  crd:
    spec:
      names:
        kind: K8sAllowedRepos
      validation:
        # Schema for the `parameters` field
        openAPIV3Schema:
          type: object
          properties:
            repos:
              type: array
              items:
                type: string
  targets:
    - target: admission.k8s.gatekeeper.sh
      rego: |
        package k8sallowedrepos

        violation[{"msg": msg}] {
          container := input.review.object.spec.containers[_]
          satisfied := [good | repo = input.parameters.repos[_] ; good = startswith(container.image, repo)]
          not any(satisfied)
          msg := sprintf("container <%v> has an invalid image repo <%v>, allowed repos are %v", [container.name, container.image, input.parameters.repos])
        }

        violation[{"msg": msg}] {
          container := input.review.object.spec.initContainers[_]
          satisfied := [good | repo = input.parameters.repos[_] ; good = startswith(container.image, repo)]
          not any(satisfied)
          msg := sprintf("container <%v> has an invalid image repo <%v>, allowed repos are %v", [container.name, container.image, input.parameters.repos])
        }
```
```yaml
apiVersion: expansion.gatekeeper.sh/v1alpha1
kind: ExpansionTemplate
metadata:
  name: expand-deployments
spec:
  applyTo:
    - groups: ["apps"]
      kinds: ["DeamonSet", "Deployment", "Job", "ReplicaSet", "ReplicationController", "StatefulSet"]
      versions: ["v1"]
  templateSource: "spec.template"
  generatedGVK:
    kind: "Pod"
    group: ""
    version: "v1"
```

Reply:
thanks! @davis-haba are you able to reproduce?

Reply:
Hey @ritazh ,

I was not able to reproduce this. The Audit yields 2 violations --one for ReplicaSet and one for the Deployment.

```   totalViolations: 2
  violations:
  - enforcementAction: deny
    group: apps
    kind: ReplicaSet
    message: '[Implied by expand-deployments] container <nginx> has an invalid image
      repo <nginx:1.14.2>, allowed repos are ["openpolicyagent"]'
    name: nginx-deployment-6c9457f8c8
    namespace: default
    version: v1
  - enforcementAction: deny
    group: apps
    kind: Deployment
    message: '[Implied by expand-deployments] container <nginx> has an invalid image
      repo <nginx:1.14.2>, allowed repos are ["openpolicyagent"]'
    name: nginx-deployment
    namespace: default
    version: v1

```

I first created the deployment, waited a little, then created the constraint template, constraint, and expansion template. The next audit cycle added the 2 aforementioned violations.

Reply:
I was able to reproduce the issue with the helm chart. Please apply the suggestions here: https://github.com/open-policy-agent/gatekeeper/pull/2062/files#r971244677

Reply:
Thanks Rita! I will be sure to test new feature flags with Helm moving forward.